include/rtio.h

```diff
@@ -0,0 +1,479 @@
+/*
+ * Copyright (c) 2019  Thomas Burdick <thomas.burdick@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_RTIO_H_
+#define ZEPHYR_INCLUDE_RTIO_H_
+/**
+ * @brief RTIO Interface
+ * @defgroup rtio_interface RTIO Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <kernel.h>
+#include <device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @file
+ * @brief Real-Time IO device API for moving bytes around fast
+ *
+ * Each Real-Time IO device provides a way to either read, write, or both
+ * blocks of memory allocated from a static pool by the application. The layout
+ * of the each block is understood implicitly by the author. Each device may
+ * provide functionality to safely encode or decode blocks of memory.
+ */
+
+/**
+ * @brief Chunk of memory with ownership and byte format versioning
+ */
+struct rtio_block {
```

Comment:
Did you think about connecting blocks as a list?
Maybe you should have a look at net_buf implementation. Those buffers are really nice to handle.

Reply:
I've more than once considered just using net_buf, I wanted a very specific allocation strategy though and I need a contiguous memory area for dma, I'll look again at using net_buf more directly here as it provides all the niceties of pushing/pulling various primitives already with it I believe

Reply:
It was just a hint. I'm not sure if it makes sense to have the ability to put the buffers into a list.

Reply:
I've changed this API so that blocks may be allocated by any allocation scheme, with a mempool implementation provided by a macro.